File: lib/tagset.ts
<!doctype tagset system "tagset.dtd">
<!-- ---------------------------------------------------------------------- -->
<!-- The contents of this file are subject to the Ricoh Source Code Public  -->
<!-- License Version 1.0 (the "License"); you may not use this file except  -->
<!-- in compliance with the License.  You may obtain a copy of the License  -->
<!-- at http://www.risource.org/RPL                                         -->
<!--                                                                        -->
<!-- Software distributed under the License is distributed on an "AS IS"    -->
<!-- basis, WITHOUT WARRANTY OF ANY KIND, either express or implied.  See   -->
<!-- the License for the specific language governing rights and limitations -->
<!-- under the License.                                                     -->
<!--                                                                        -->
<!-- This code was initially developed by Ricoh Silicon Valley, Inc.        -->
<!-- Portions created by Ricoh Silicon Valley, Inc. are                     -->
<!-- Copyright (C) 1995-1999.  All Rights Reserved.                         -->
<!--                                                                        -->
<!-- Contributor(s): steve@rii.ricoh.com pgage@rii.ricoh.com                -->
<!-- ---------------------------------------------------------------------- -->

<tagset name="tagset" parent="HTML">
<title>Tagset for Tagset Definition Files</title>
<cvs-id>$Id: tagset.ts,v 1.2 2001-01-11 23:37:01 steve Exp $</cvs-id>

<h2>Introduction:</h2>

<p> This file contains the <a href="#XHML">eXtended HTML (XHML)</a> definition
    for the <code>tagset</code> <a href="#tagset">tagset</a>, or in other
    words, the formal definition of the language in which tagsets are defined.
    (Yes, this is a <a href="#circular">circular definition</a>.)  You are
    probably reading it in HTML form, as <a href="tagset.html"
    ><code>tagset.html</code></a>; this was derived automatically from the
    XHML original file, <a href="tagset.ts"><code>tagset.ts</code></a>.
</p>

<p> A tagset definition file such as this is essentially a mapping into
    XHML of an SGML Document Type Definition (DTD), and in fact can be
    automatically processed into one.  The associated document-processing
    semantics are, however, lost in the translation process, along with some
    of the associated documentation, which is represented in HTML.  The
    developers of this system hope that you will find the resulting documents
    easier to read than the equivalent DTD's. 
</p>

<p> (Another way of looking at a tagset definition file is as an HTML file
    that documents a markup language, augmented with a handful of XML
    constructs that actually <em>define</em> the language.  Because both
    the formal and informal descriptions are mixed in the same file, it is
    easy to derive some of the documentation directly from the formal
    components, thus ensuring its correctness.  This allows the developer to
    confine the <em>informal</em> parts of the documentation to descriptive
    material, without having to worry about keeping two equivalent versions of
    every definition in sync.)
</p>

<h3>How to Read Tagset Descriptions</h3>

<p> A tagset definition file consists of a sequence of <tag>define</tag>
    elements, some of which are nested inside others.  In particular, the
    outermost <tag>define</tag> elements define SGML ``<a
    href="#elements">elements</a>'' and ``<a href="#entities">entities</a>''.
    Nested inside each element definition are the definitions of its
    ``attributes''.  Any definition can also contain documentation.
</p>

<p> When converted to HTML for documentation, the nesting of attribute
    definitions inside of element definitions is rendered as indentation.
    Documentation is nested one more level and typeset in italics.  The
    <code>&lt;/define&gt;</code>, <code>&lt;doc&gt;</code>, and
    <code>&lt;/doc&gt;</code> tags are omitted.
</p>

<p> In some cases, certain elements are only meaningful inside of other
    elements; for example, <tag>then</tag> elements only occur inside of
    <tag>if</tag> elements.  By convention, the definitions of these elements
    follow that of their parent in a subsection containing an unordered
    (bulleted) list. 
</p>

<p> Other constructs, mainly involving attributes of <tag>define</tag>, are
    documented below at the point where they are first used. 
</p>

<h3>How Resources are Located and Loaded</h3>

<p> Java's resource mechanism allows arbitrary data files to exist in the same
    namespace as classes.  They can be shipped around in JAR files, downloaded
    from the Net, and (in most cases) obtained from directories in the
    <code>CLASSPATH</code>.  In particular, this means that common tagsets can
    be defined in the same directory as the classes of the
    <code>org.risource.dps.tagset</code> package.
</p>

<p> Tagsets and other resources (including external entities) can exist either
    as Java resources or as files located relative to the document being
    processed.  They can also be located in the DPS, in an internal namespace.
</p>

<p> When a tagset is specified in a document's DTD, it will be a system
    identifier with the extension ``<code>.dtd</code>''.  This is stripped off
    to facilitate searching for a corresponding ``<code>.ts</code>'' file
    instead.  A tagset specified on a command line takes precedence over one
    specified in the document's <code>doctype</code> declaration.  This allows
    documents to be processed with tagsets other than the one they were
    originally written for.
</p>

<p> The search order for tagsets is as follows:
</p>

<ol>
  <li> First, the namespace is identified.  An identifier followed by a colon
       indicates an explicit namespace.  Otherwise, if the name contains
       slashes it is considered to be a file located relative to the current
       document; otherwise if it contains periods it is considered to be a
       resource.  Tagsets with neither are looked for first in the current
       directory (the explicit namespace <code>file:</code>), then in the
       package <code>org.risource.dps.tagset</code> in the explicit namespace
       <code>DPS:</code>.
  </li>

  <li> If the namespace is specified explicitly, it performs the lookup.
  </li>

  <li> The tagset's name is first looked up along the internal processing
       context chain defined by the <code>topProcessor</code> links in
       <code>Context</code> objects.  
  </li>

  <li> Class <code><em>tsname</em>.class</code> is searched for provided the
       name could be in the package/resource namespace.  If it exists and
       implements the <code>Tagset</code> interface, it is loaded and
       instantiated.  
  </li>

  <li> <code><em>tsname</em>.ts</code> is searched for, first in the current
       directory, then as a resource, to identify the correct namespace.
  </li>

  <li> <code><em>tsname</em>.obj</code> is loaded as a serialized
       object, if it exists in the same location and is newer than 
       <code><em>tsname</em>.ts</code>
  </li>
       
  <li> <code><em>tsname</em>.tss</code> is processed using the minimal
       bootstrap tagset containing <tag>tagset</tag>, <tag>define</tag>, and
       their sub-elements, if it exists in the same location and is newer than
       <code><em>tsname</em>.ts</code>.  This is assumed to be a tagset
       ``stripped'' of its documentation.
  </li>

  <li> <code><em>tsname</em>.ts</code> is processed using the
       <code>tagset</code> tagset. 
  </li>
</ol>

<p> Once a tagset is loaded, its name is added to the appropriate namespace. 
</p>

<p> The search order for entities is similar, except that entities cannot be
    represented as classes, and external entity names <em>are</em> expected to
    have an extension. 
</p>

<p> While processing a tagset's start tag, the tagset mentioned in its
    <code>context</code> is also loaded, followed by any tagsets mentioned in
    its <code>include</code> attribute.  Tagsets and entities mentioned in
    resources are not looked for relative to the current document, but only
    relative to the current resource.
</p>


<h2>Definition Elements</h2>

<h3>Define and its components</h3>

<define element=define handler no-text default-content=value>
  <doc> Defines an element, attribute, entity, or word.  It is meaningful for
	for <tag>define</tag> to occur outside of a <tag>namespace</tag> or
	<tag>tagset</tag> element because there is always a ``current''
	namespace and tagset in effect.
  </doc>
  <define attribute=element optional>
    <doc> Specifies that an element (tag) is being defined.  The value of the
	  attribute is the tagname of the element being defined.  If the
	  <code>handler</code> attribute or the <code><tag>action</tag></code>
	  sub-element is present, the element will be active.  If a
	  <code>value</code> attribute or sub-element is present, the element
	  will be passively replaced by its value when the document is
	  processed.
    </doc>
  </define>
  <define attribute=attribute optional>
    <doc> Specifies that an attribute is being defined.    The value of the
	  attribute is the name of the attribute being defined.  If the
	  <code>handler</code> attribute or the <code><tag>action</tag></code>
	  sub-element is present, the attribute will be active.  If a
	  <code>value</code> attribute or sub-element is present, the
	  attribute will be passively replaced by its value unless a value is
	  explicitly specified in the tag where the attribute occurs.

	  <p>It is usual for attributes to be defined <em>inside</em> an
	  element's definition.</p>
    </doc>
  </define>
  <define attribute=entity optional>
    <doc> Specifies that an entity is being defined.    The value of the
	  attribute is the name of the entity being defined.  If the
	  <code>handler</code> attribute or the <code><tag>action</tag></code>
	  sub-element is present, the entity will be active.  If a
	  <code>value</code> attribute or sub-element is present, the entity
	  will be passively replaced by its value when it is referenced in a
	  document.
    </doc>
  </define>
  <define attribute=notation optional>
    <doc> Specifies that a notation is being defined.    The value of the
	  attribute is the name of the notation being defined.  The associated
	  value, if any, should be the MIME type of the data.  A ``notation''
	  is the identifier used in a DTD to describe the data in an unparsed
	  external entity.
    </doc>
  </define>

  <doc> <strong>General modifiers:</strong>
  </doc>

  <define attribute=handler optional>
    <doc> Specifies the action handler class for the node being defined.

	  <p> If no value is specified, the handler class name is assumed to
	  be the same as the element's tag, possibly with <code>Handler</code>
	  appended.  If the value contains periods, it is assumed to be a
	  complete Java class name and is used unmodified.</p>

	  <p><strong>Note</strong> that if the Node being defined is an
	  <em>attribute</em>, things get a little complicated.  What we would
	  really like to happen is that the parser selects a handler for the
	  containing <em>element</em> based on the presence of a ``handled''
	  attribute.  Probably only one such attribute should be permitted.</p>
    </doc>
    <note author=steve>
	=== the implementation needs to give an error for missing handlers!
    </note>
  </define>

  <doc> <strong>Modifiers for <code><tag>define element</tag></code>:</strong>
	<blockquote><em>
	  The following attributes are only meaningful when defining an
	  Element.  It is impossible to represent this constraint in SGML.
	</em></blockquote>
  </doc>

  <define attribute=quoted optional>
    <doc> Indicates that the content of the element being defined should be
	  parsed, but not expanded before invoking the action.
    </doc>
  </define>
  <define attribute=literal optional>
    <doc> Indicates that the content of the element is unparsed (#CDATA).
    </doc>
  </define>
  <define attribute=empty optional>
    <doc> Indicates that the element being defined will never have any
	  content. 
    </doc>
  </define>

  <doc> <strong>Modifiers for <code><tag>define attribute</tag></code>:</strong>
  </doc>

  <define attribute=optional optional>
    <doc> Only meaningful for attributes.  Specifies that the attribute is
	  optional (optional).
    </doc>
  </define>
  <define attribute=required optional>
    <doc> Only meaningful for attributes.  Specifies that the attribute is
	  required.
    </doc>
  </define>
  <define attribute=fixed optional>
    <doc> Only meaningful for attributes.  Specifies that the attribute's
	  value is invariant (and must be given in its definition).
    </doc>
  </define>

  <doc> <strong>Modifiers for <code><tag>define entity</tag></code>:</strong>
  </doc>

  <define attribute=system optional>
    <doc> The value of this attribute is the ``system identifier'' (URI) of
	  an ``external entity''.  Usually it will be a filename relative to
	  the document containing the definition being processed.
    </doc>
  </define>
  <define attribute=public optional>
    <doc> The value of this attribute is the ``public identifier'' of an
	  external entity.  In a DTD, an alternative system identifier must be
	  provided; that should be specified via the <code>system</code>
	  attribute. 
    </doc>
  </define>
  <define attribute=NDATA optional>
    <doc> This specifies that the entity contains non-parsed data; the value
	  specifies the name of the data's <em>notation</em>.
    </doc>
  </define>
  <define attribute=retain optional>
    <doc> This specifies that the a reference to the entity should be replaced
	  by its value when conversion to a string is desired, but otherwise
	  should be retained in the tree and passed through to the output.
	  This is used, e.g., for the predefined character entities.
    </doc>
  </define>

</define>

<dl>
  <dt> Note:</dt>
  <dd> The <code>no-text</code> attribute specifies that the element, in this
       case <tag>define</tag>, does not contain text.  All whitespace in its
       content is marked ignorable.  The <code>default-content=</code>
       attribute specifies the tag to use to ``wrap'' any content which is not
       one of the defined children.
  </dd>       
</dl>

<h4>Sub-elements of <tag>define</tag></h4>
<dl>
  <dt> Note:</dt>
  <dd> The use of <code>parent=</code> specifies that these elements only
       occur inside the given parent element; in this case,
       <code>define</code>.  The value of the <code>parent</code> attribute is
       actually a list which is appended to with each use, allowing the DTD to
       be incrementally extended.

       <p>The use of <code>parent</code> greatly simplifies the construction of
       content models and the parser.  An element with a parent implicitly
       terminates any unclosed elements between it and its innermost parent.</p>
  </dd>       
</dl>

<ul>
  <li> <define element=value parent=define>
         <doc> The <tag>value</tag> sub-element defines a value for the node
	       being defined.  The node will be replaced by its value whenever
	       it appears in a document being processed.
         </doc>
       </define>
  </li>
  <li> <define element=action parent=define quoted>
         <doc> The <tag>action</tag> sub-element defines an action for the
	       node being defined.  Note that it is possible for a node to
	       have both an action and a value.
         </doc>
       </define>
  </li>
</ul>

<h3>Documentation Elements</h3>

<ul>
  <li> <define element="doc">
         <doc> This sub-element contains documentation for the node being
	       defined.  It may be either retained or stripped out depending
	       on how the enclosing namespace is being processed. 
         </doc>
       </define>
  </li>

  <li> <define element="note">
         <doc> This sub-element contains attributed annotation for the node
	       being defined.  It may be either retained or stripped out
	       depending on how the enclosing namespace is being processed.
         </doc>
         <define attribute="author" optional>
       	   <doc> The value of this attribute should be the author's initials,
		 login name, or e-mail address.
           </doc>
         </define>
       </define>
  </li>
</ul>

<h3>Tagset and Namespace</h3>

<define element=tagset handler>
  <doc> This element defines a ``<em>tagset</em>'' -- roughly the equivalent
	of a DTD or database schema.  Having an XML representation allows
	tagsets to be processed using ``normal'' methods, rather than devising
	special machinery for parsing and processing DTD's.
  </doc>
  <note author=steve>
	Eventually it will be possible to process XML tagsets into DTD's.
  </note>
  <note author=steve>
	Eventually we must make it possible to ``include'' one tagset
	inside another by using <tag>extract</tag>.
  </note>

  <define attribute=context optional>
    <doc> The <code>context</code> attribute specifies the tagset in which
	  names not defined in the current tagset will be looked up.  It is
	  effectively <em>included</em> in the tagset being defined.
    </doc>
  </define>
  <define attribute=include optional>
    <doc> The <code>include</code> attribute specifies a list of tagsets,
	  by name, whose contents are to be <em>copied into</em> the current
	  <tag>tagset</tag>.  This is different from <tag>context</tag>, which
	  effectively includes by reference.
    </doc>
  </define>
  <define attribute=recursive optional>
    <doc> If present, this attribute indicates that elements defined in the
	  tagset can be used in the definitions of other elements.  The
	  default is for definitions to be in terms of elements and entities
	  defined in the <em>enclosing document's</em> tagset.  This allows
	  restricted tagsets to be defined.
    </doc>
  </define>
</define>

<define element=namespace handler>
  <doc> This defines a namespace for entities. 
  </doc>
  <define attribute=name optional>
    <doc> Note that the <code>name</code> attribute is optional; it is
	  perfectly meaningful to have an anonymous <tag>namespace</tag>.
    </doc>
  </define>
  <define attribute=context optional>
    <doc> The <code>context</code> attribute specifies the namespace in which
	  names not defined in the current tagset will be looked up.  It is
	  effectively <em>included</em> in the tagset being defined.  If no
	  value is specified, the innermost namespace is assumed.
    </doc>
  </define>
  <define attribute=include optional>
    <doc> The <code>include</code> attribute specifies a list of namespaces,
	  by name, whose contents are to be <em>copied into</em> the current
	  <tag>namespace</tag>.  This is different from <tag>context</tag>,
	  which effectively includes by reference.
    </doc>
  </define>
</define>


<h2>Formal documentation comment</h2>

<doc> This tagset defines the syntax and semantics of files that define
      tagsets.  It exists only for documentation purposes; the actual
      machinery has to be present during initialization.
</doc>

<h2>Notes and Glossary</h2>

<dl>
  <dt> <a name="XXML">XXML</a> (eXtended, or eXecutable, XML)</dt>
  <dd> XML (Extensible Markup Language) augmented with semantic descriptions,
       also XML, that specify <em>actions</em> to be taken by a suitable
       document processor such as the DPS.  XML is an SGML ``application''.  A
       document in XXML is called an ``active document''.
  </dd>

  <dt> <a name="XHML">XHML</a> (eXtended Hypertext Markup Language)</dt>
  <dd> HTML (HyperText Markup Language) extended with <a href="#XXML">XXML</a>
       constructs.  Because HTML includes SGML constructs that are not
       permitted in XML, XXML strictly speaking does not conform to the XML
       specification.  It does, however, conform to the <em>SGML</em>
       specification.
  </dd>

  <dt> <a name="tagset">Tagset</a></dt>
  <dd> A Tagset is the XHML representation of an SGML Document Type
       Definition (DTD), augmented with additional semantic information that
       specifies the actions to be taken by a document processing system when
       it encounters certain ``active'' constructs in the document.
  </dd>

  <dt> <a name="circular">Circular Definition</a></dt>
  <dd> See ``Circular Definition'' -- something defined in terms of itself.
       The key to unwinding the circularity involved in the Tagset called
       ``<code>tagset</code>'' (which defines the syntax and semantics of
       tagset files) is to observe that the definitions <em>really</em>
       required to get started, namely <tag>tagset</tag>, <tag>define</tag>,
       and their components, are ``compiled in'' to the document processor.
       These, plus the fact that XML and almost all of HTML can be parsed with
       the default XML syntax, are sufficient for reading in stripped-down
       versions of the HTML and Basic tagsets.
  </dd>

  <dt> <a name="elements">Elements</a></dt>
  <dd> those SGML constructs that look like:
       <pre>
       &lt;<em>tagname</em> <em>attribute</em>=<em>value</em>&gt;
       <em>... content ...</em>
       &lt;/<em>tagname</em>&gt;
       </pre>

       The construct in angle brackets in at the start of the element is
       called the ``start tag'', and the construct at the end is called the
       ``end tag''.  An Element which can  never have any content is called
       ``empty'', and its end tag can be omitted.

       <p>In the Document Processing System, elements can be defined to have
       associated ``actions'' that operate like macros or functions.  These
       elements are called ``active''.</p>
  </dd>

  <dt> <a name="entities">Entities</a></dt>
  <dd> Those SGML constructs that look like
       ``<code>&amp;<em>name</em>;</code>'' and which represent a named piece
       of content to be substituted into the document.  In the Document
       Processing System, entities can be used similarly to variables in
       programming languages.
  </dd>
</dl>

<hr>
<b>Copyright &copy; 1998 Ricoh Innovations, Inc.</b><br>
<b>$Id: tagset.ts,v 1.2 2001-01-11 23:37:01 steve Exp $</b><br>
</tagset>

File: src/java/org/risource/dps/Tagset.java
////// Tagset.java: Node Handler Lookup Table interface
//	$Id: Tagset.java,v 1.10 1999-08-20 00:00:58 steve Exp $

/*****************************************************************************
 * The contents of this file are subject to the Ricoh Source Code Public
 * License Version 1.0 (the "License"); you may not use this file except in
 * compliance with the License.  You may obtain a copy of the License at
 * http://www.risource.org/RPL
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied.  See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * This code was initially developed by Ricoh Silicon Valley, Inc.  Portions
 * created by Ricoh Silicon Valley, Inc. are Copyright (C) 1995-1999.  All
 * Rights Reserved.
 *
 * Contributor(s):
 *
 ***************************************************************************** 
*/


package org.risource.dps;

import java.util.Enumeration;
import java.io.File;

import org.risource.dps.active.*;
import org.w3c.dom.NamedNodeMap;

/**
 * The interface for a Tagset -- a lookup table for syntax. <p>
 *
 *	A Node's Handler provides all of the necessary syntactic and semantic
 *	information required for parsing, processing, and presenting the Node.
 *	A Tagset can be regarded as either a lookup table for syntactic
 *	information, or as a node factory for the documents so described.  <p>
 *
 *	Note that a Tagset can be used to construct either generic DOM Node's,
 *	or DPS ActiveNode's.  A Parser is free to use either.  In the current
 *	implementation, however, <em>all</em> documents produced in the DPS
 *	are Active parse trees, so the default is for the ``generic'' nodes
 *	produced by the standard tagsets are identical to the active ones.
 *	The only difference is the return type.  The specialized Output
 *	ToDocument should be used for conversion.  <p>
 *
 *	Note that this interface says little about the implementation.
 *	It is expected, however, that any practical implementation of
 *	Tagset will also be a Node, so that tagsets can be read and
 *	stored as documents or (better) DTD's.  <p>
 *
 * === 	need encoders/decoders for character entities, URLs, etc.
 *
 * @version $Id: Tagset.java,v 1.10 1999-08-20 00:00:58 steve Exp $
 * @author steve@rsv.ricoh.com
 *
 * @see org.risource.dps.Processor
 * @see org.risource.dps.Active
 * @see org.risource.dps.active.ActiveNode
 * @see org.risource.dps.Input
 * @see org.risource.dps.Output
 * @see org.risource.dps.output.ToDocument
 */

public interface Tagset  {

  /************************************************************************
  ** Context:
  ************************************************************************/

  /** Returns a Tagset which will handle defaults. 
   *	Note that it may or may not be used by the various lookup
   *	operations; it will usually be more efficient to duplicate the
   *	entries of the context.  However, lightweight implementations
   *	that define only a small number of tags may use it.
   */
  public Tagset getContext();

  /** Include definitions from a given tagset. 
   */
  public void include(Tagset ts);

  /** Include definitions, defaults, and other parameters from a given tagset. 
   */
  public void setParent(Tagset ts);

  /** Get the tagset's name. */
  public String getName();


  /************************************************************************
  ** Entity Bindings:
  ************************************************************************/

  /** Return a namespace with a given name.  Returns the entity table 
   *	associated with the tagset (possibly in the context chain) having 
   *	the given name.
   */
  public Namespace getNamespace(String name);

  /** Obtain the current Entity bindings. */
  public EntityTable getEntities();

  /** Get the binding (Entity node) of an entity, given its name. 
   * @return <code>null</code> if the entity is undefined.
   */
  public ActiveNode getBinding(String name);

  /** Bind an Entity.
   */
  public void setBinding(String name, ActiveNode ent);


  /************************************************************************
  ** Lookup Operations:
  ************************************************************************/

  /** Called during parsing to return a suitable Handler for a given
   *	tagname.  Returns <code>getHandlerForType(Node.ELEMENT_NODE)</code> 
   *	as a default.
   */
  public Handler getHandlerForTag(String tagname);

  public void setHandlerForTag(String tagname, Handler newHandler);

  /** Called during parsing to return a suitable Handler for a generic
   *	Node, given the Node's type.
   */
  public Handler getHandlerForType(short nodeType);

  public void setHandlerForType(short nodeType, Handler newHandler);

  /** Called during parsing to determine whether a Handler exists for a
   *	given attribute name.  Returns <code>null</code> as a default.
   */
  public Handler getHandlerForAttr(String attrName);

  public void setHandlerForAttr(String attrName, Handler newHandler);

  /************************************************************************
  ** Status:
  ************************************************************************/

   /** Test whether the Tagset is ``locked.''
   *
   *	A locked Tagset must be extended by creating a new Tagset with
   *	the locked Tagset as its context.
   */
  public boolean isLocked();

  /** Change the lock status. */
  public void setLocked(boolean value);

  /** Test whether the Tagset is up to date.  */
  public boolean upToDate();

  /** Set the Tagset's source File for timestamp check. */
  public void setSourceFile(File f);

  /************************************************************************
  ** Parsing Operations:
  ************************************************************************/

  /** Return a Parser suitable for parsing a character stream
   *	according to the Tagset.  The Parser may (and probably will)
   *	know the Tagset's actual implementation class, so it can use
   *	specialized operations not described in the Tagset interface.
   */
  public Parser createParser();

  /** Creates an ActiveAttrList from a NamedNodeMap. */
  public ActiveAttrList createActiveAttrs(NamedNodeMap attrs);

  /** Creates an ActiveElement suitable for use in any document described
   *	by this Tagset.
   */
  public ActiveElement createActiveElement(String tagname,
					   ActiveAttrList attributes,
					   boolean hasEmptyDelim);

  /** Creates an ActiveNode of arbitrary type with name and (optional) data.
   */
  public ActiveNode createActiveNode(short nodeType, String name, String data);

  /** Creates an ActiveNode of arbitrary type with name and content.
   */
  public ActiveNode createActiveNode(short nodeType, String name,
				     ActiveNodeList data);

  /** Creates an ActiveAttribute node with name and value.
   */
  public ActiveAttr createActiveAttr(String name, ActiveNodeList value);

  /** Creates an ActiveEntity node with name and value.
   */
  public ActiveEntity createActiveEntity(String name, ActiveNodeList value);

  /** Creates an ActiveText node.  Otherwise identical to createText.
   */
  public ActiveText createActiveText(String text, boolean isIgnorable);

  /************************************************************************
  ** Syntactic Information:
  ************************************************************************/

  /** Does this Tagset treat uppercase and lowercase tagnames the same?
   */
  public boolean caseFoldTagnames();

  /** Convert a tagname to the cannonical case. */
  public String cannonizeTagname(String name);

  /** Does this Tagset treat uppercase and lowercase attribute names 
   *	the same?
   */
  public boolean caseFoldAttributes();

  /** Set the case folding parameters */
  public void setCaseFolding(boolean forTags, boolean forAttrs);


  /** Convert an attribute name to the cannonical case. */
  public String cannonizeAttribute(String name);


  /** Return the tag of the paragraph element, implicitly started
   *	when text appears inside an element that should not contain it. 
   */
  public String paragraphElementTag();

  /** Return the Tagset's DTD.  In some implementations this may be
   *	the Tagset itself.
   */
  //  public DocumentType getDocumentType();


  /************************************************************************
  ** Documentation Operations:
  ************************************************************************/

  /** Returns an Enumeration of the element names defined in this
   *	table.  Note that there is no good way to get the handlers for
   *	Node types other than Element unless the implementation gives
   *	them distinctive, generated names.
   */
  public Enumeration handlerNames();

  /** Returns an Enumeration of the element names defined in this table and
   *	its context, in order of definition (most recent last). */
  public Enumeration allHandlerNames();

  /************************************************************************
  ** Convenience Functions:
  ************************************************************************/

  /** Convenience function to define a tag with a given syntax. */
  public Handler defTag(String tag, String notIn, String parents, int syntax,
			String cname, ActiveNodeList content);

}
